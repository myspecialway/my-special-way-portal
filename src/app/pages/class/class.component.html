<div class="main-content msw-class-container" data-test-id="classes-page">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">כיתות</span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button data-test-id="add-class-button" mat-raised-button color="primary" [routerLink]="['/class', '_new_']">הוסף כיתה חדשה</button>
      </div>
    </div>
    <div class="row msw-table-container">

      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table"
                   matSort matSortActive="name" matSortDisableClear matSortDirection="asc">

          <!-- Number Column -->
          <ng-container matColumnDef="classname">
            <mat-header-cell *matHeaderCellDef>שם כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="navigate(row._id)" class="class-name">{{ row.name }}</mat-cell>
          </ng-container>

          <!-- Title Column -->
          <ng-container matColumnDef="level">
            <mat-header-cell *matHeaderCellDef>כיתה</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="navigate(row._id)">{{ scheduleService.grades[row.grade] }}</mat-cell>
          </ng-container>

          <!-- State Column-->
          <ng-container matColumnDef="deleteClass">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button class="delete-row-button" [attr.data-test-id]="'delete-class-button-' + row.name" mat-icon-button color="primary" (click)="deleteClass(row._id, row.name, row.level, row.students.length)"><i class="material-icons"> delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns"></mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>


<!-- Copyright 2018 Google Inc. All Rights Reserved.
    Use of this source code is governed by an MIT-style license that
    can be found in the LICENSE file at http://angular.io/license -->
