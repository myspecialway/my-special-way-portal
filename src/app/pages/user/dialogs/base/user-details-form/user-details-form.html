<form #formControl="ngForm" [formGroup]="form" novalidate (submit)="submitForm()">
  <mat-dialog-content>
    <div class="msw-div-column-right">
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <input tabindex="1" matInput #input formControlName="firstname" placeholder="שם פרטי" [(ngModel)]="data.firstname"
            autocomplete='given-name' name="firstname" required>
          <button type="button" mat-button *ngIf="data.firstname" matSuffix mat-icon-button aria-label="Clear"
            (click)="data.firstname=''">
            <mat-icon>close</mat-icon>
          </button>
          <mat-error *ngIf="formControl.invalid">שם פרטי הינו שדה חובה</mat-error>
        </mat-form-field>
      </div>
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <input matInput #input tabindex="3" formControlName="username" placeholder="שם משתמש" [(ngModel)]="data.username"
            autocomplete="username" name="username" required>
          <button type="button" mat-button *ngIf="data.username" matSuffix mat-icon-button aria-label="Clear"
            (click)="data.username=''">
            <mat-icon>close</mat-icon>
          </button>
          <mat-error *ngIf="userNameFormControl.hasError('required')">שם משתמש הינו שדה חובה</mat-error>
          <mat-error *ngIf="userNameFormControl.hasError('minlength') || userNameFormControl.hasError('pattern')">שם
            משתמש חייב להיות לפחות חמישה תווים באנגלית</mat-error>
        </mat-form-field>
      </div>
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <mat-select tabindex="5" placeholder="סוג משתמש" formControlName="role" [(ngModel)]="data.role" name="role"
            (valueChange)="onUserTypeChange($event)" autocomplete='role' required>
            <mat-option disabled>בחר</mat-option>
            <mat-option *ngFor="let _role of roles" [value]="_role">
              {{userRoleEnum[_role]}}
            </mat-option>
          </mat-select>
          <mat-error *ngIf="selectUserType.hasError('required')">אנא בחר סוג משתמש</mat-error>
        </mat-form-field>
      </div>
    </div>
    <div class="msw-div-column-left">
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <input tabindex="2" matInput #input formControlName="lastname" placeholder="שם משפחה" [(ngModel)]="data.lastname"
            autocomplete='family-name' name="lastname" required>
          <button tabindex="0" type="button" mat-button *ngIf="data.lastname" matSuffix mat-icon-button aria-label="Clear"
            (click)="data.lastname=''">
            <mat-icon>close</mat-icon>
          </button>
          <mat-error *ngIf="formControl.invalid">שם משפחה הינו שדה חובה</mat-error>
        </mat-form-field>
      </div>
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <input tabindex="4" matInput #input formControlName="email" placeholder="אימייל" [(ngModel)]="data.email" name="email"
            autocomplete='email' required email>
          <button type="button" mat-button *ngIf="data.email" matSuffix mat-icon-button aria-label="Clear"
            (click)="data.email=''">
            <mat-icon>close</mat-icon>
          </button>
          <mat-error *ngIf="EmailFormControl.hasError('required')">אימייל הינו שדה חובה</mat-error>
          <mat-error *ngIf="EmailFormControl.hasError('email')">אימייל לא חוקי</mat-error>
        </mat-form-field>
      </div>
      <div class="msw-form-row">
        <mat-form-field color="primary">
          <mat-select tabindex="6" placeholder="כיתה" formControlName="class" [(ngModel)]="data.class" name="class" [compareWith]="compareClass">
           <mat-option disabled>בחר</mat-option>
           <mat-option *ngFor="let _class of classes" [value]="_class">
              {{_class.name}}
            </mat-option>
          </mat-select>
          <mat-error *ngIf="selectGrade.hasError('required')">כיתה הינו שדה חובה</mat-error>
        </mat-form-field>
      </div>
    </div>
  </mat-dialog-content>
  <div mat-dialog-actions>
    <button type="button" mat-button class="msw-button-cancel" tabindex="-1" (click)="close()">ביטול</button>
    <button type="submit" mat-button class="msw-button-submit" [disabled]="!formControl.valid">{{options.submitButtonLabel}}</button>
  </div>
</form>
