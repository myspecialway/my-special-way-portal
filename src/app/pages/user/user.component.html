<div data-test-id="users-page" class="main-content">
  <div class="container-fluid">
    <div class="header-menu-container msw-table-tools-header">
      <div class="header-table-count">
        <div>
          <span class="header-table-count-label">
          {{'USERS.USERS_TITLE' | translate}}
          </span>
          <span class="header-table-count-number">({{dataSource.data.length}})</span>
        </div>
      </div>
      <div class="example-header msw-table-tools-header">
        <button mat-raised-button color="primary" (click)="addNewUser()" title="{{'USERS.USERS_INSERT' | translate}}">
          <i class="material-icons">
           add
          </i> {{'USERS.USERS_INSERT' | translate}}
       
        </button>
      </div>
    </div>
    <div class="row msw-table-container">
      <div class="col-lg-12 col-md-12 col-sm-12">
        <mat-table #table [dataSource]="dataSource" class="msw-table"
                   matSort matSortActive="firstName" matSortDisableClear matSortDirection="asc">

          <!-- ID Column -->
          <ng-container matColumnDef="name">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_NAME' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)"> {{row.firstname}} {{ row.lastname }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="class">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_CLASS' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{(row.class?.name) || '' }}</mat-cell>
          </ng-container>

          <!-- Progress Column -->
          <ng-container matColumnDef="username">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_USERNAME' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{row.username}}</mat-cell>
          </ng-container>


          <!-- Name Column -->
          <ng-container matColumnDef="type">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_USERROLE' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{toHebrew(row.role)}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="email">
            <mat-header-cell *matHeaderCellDef mat-sort-header> {{'USERS.USERS_EMAIL' | translate}}</mat-header-cell>
            <mat-cell *matCellDef="let row" (click)="updateUser(row)">{{row.email}}</mat-cell>
          </ng-container>

          <ng-container matColumnDef="recoverpassword">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button class="recover-password-button" mat-icon-button color="primary"> {{'USERS.USERS_RECOVERPASS' | translate}}</button>
            </mat-cell>
          </ng-container>

          <ng-container matColumnDef="deleteUser">
            <mat-header-cell *matHeaderCellDef></mat-header-cell>
            <mat-cell *matCellDef="let row">
              <button class="delete-button" mat-icon-button color="primary" (click)="deleteUser(row)" title="{{'COMMONS.DELETE' | translate}}"><i class="material-icons">delete </i></button>
            </mat-cell>
          </ng-container>

          <mat-header-row *matHeaderRowDef="displayedColumns"></mat-header-row>
          <mat-row *matRowDef="let row; columns: displayedColumns;">
        </mat-row>
        </mat-table>
      </div>
    </div>
  </div>
</div>
