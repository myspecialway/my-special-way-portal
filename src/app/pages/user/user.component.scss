@import '../../../assets/scss/msw-table';

.header-table-count {
  width: 164px;
  height: 38px;
  font-size: 32px;
  font-weight: normal;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  letter-spacing: 0.6px;
  text-align: right;
  color: #222222;
}
.header-table-count-number {
  font-size: 18px;
}

.header-table-count-label {
  padding-left: 10px;
}

.header-menu-container {
  padding-bottom: 29px;
}

.recover-password-button {
  text-decoration: underline;
  display: none;
}

.mat-row:hover {
  .recover-password-button {
    display: block;
  }
}
