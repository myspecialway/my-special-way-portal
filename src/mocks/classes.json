[
  {
    "_id": "5b0116f8c1d99ef1b8e73245",
    "index": 0,
    "grade": "A",
    "number": 1,
    "description": "Tiltan",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f870a67d3a47e9f235",
    "index": 1,
    "grade": "A",
    "number": 2,
    "description": "Tapuz",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f82074861f963642f3",
    "index": 2,
    "grade": "A",
    "number": 3,
    "description": "Tapuch",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8b9e91d54d315372c",
    "index": 3,
    "grade": "A",
    "number": 4,
    "description": "Tut",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8895af89de5563236",
    "index": 4,
    "grade": "B",
    "number": 1,
    "description": "Dafna",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f870788b1e9f99b814",
    "index": 5,
    "grade": "B",
    "number": 2,
    "description": "Dekel",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f896b61b3a5c135a1d",
    "index": 6,
    "grade": "B",
    "number": 3,
    "description": "Duvdevan",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f855ecf6fd19ae77fa",
    "index": 7,
    "grade": "B",
    "number": 4,
    "description": "Dolev",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 1,
    "description": "Egoz",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 2,
    "description": "Agas",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 3,
    "description": "Erez",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 4,
    "description": "Oren",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f814169898640db32f",
    "index": 8,
    "grade": "C",
    "number": 5,
    "description": "Alon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f87554b3534cfde9b7",
    "index": 9,
    "grade": "D",
    "number": 1,
    "description": "Shaked",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8cc498da37f361b2f",
    "index": 10,
    "grade": "D",
    "number": 2,
    "description": "Gefen",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f832a6de313076c3b1",
    "index": 11,
    "grade": "D",
    "number": 3,
    "description": "Shezif",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f832a6de313076c3b1",
    "index": 11,
    "grade": "D",
    "number": 4,
    "description": "Shesek",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f841bc308bff51f0b8",
    "index": 12,
    "grade": "E",
    "number": 1,
    "description": "Rimon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8897ec61406fc591f",
    "index": 13,
    "grade": "E",
    "number": 2,
    "description": "Teena",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f867a266b544055fc3",
    "index": 14,
    "grade": "E",
    "number": 3,
    "description": "Zait",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f846bd827beedec579",
    "index": 15,
    "grade": "F",
    "number": 1,
    "description": "Banana",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8ab446a3271359472",
    "index": 16,
    "grade": "F",
    "number": 2,
    "description": "Balon",
    "num_of_students": 20
  },
  {
    "_id": "5b0116f8f0319d311adbd5b8",
    "index": 17,
    "grade": "F",
    "number": 3,
    "description": "Bait",
    "num_of_students": 20
  }
]
