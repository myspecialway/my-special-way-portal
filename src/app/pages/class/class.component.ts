import { Component, OnInit, ElementRef, ViewChild } from '@angular/core';
import 'rxjs/add/observable/merge';
import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/distinctUntilChanged';
import { MatTableDataSource, MatSort, MatDialog } from '@angular/material';
import { ClassService } from './services/class.graphql.service';
import { Class } from '../../models/class.model';
import { ScheduleService } from '../../services/schedule/schedule.service';
import { MSWSnackbar } from '../../services/msw-snackbar/msw-snackbar.service';
import { DeleteClassDialogComponent } from './dialogs/delete/delete-class.dialog';
import { Router } from '@angular/router';

@Component({
  selector: 'app-grade',
  templateUrl: './class.component.html',
  styleUrls: ['./class.component.scss'],
})
export class ClassComponent implements OnInit {
  displayedColumns = ['classname', 'level', 'deleteClass'];
  dataSource = new MatTableDataSource<Class>();
  resultsLength = 0;

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  constructor(
    private classService: ClassService,
    public dialog: MatDialog,
    public scheduleService: ScheduleService,
    private mswSnackbar: MSWSnackbar,
    private router: Router,
  ) {}

  async ngOnInit() {
    await this.populateDatasource();
    this.dataSource.sort = this.sort;
  }

  private async populateDatasource() {
    try {
      this.classService.getAllClasses().subscribe((classes) => {
        this.dataSource.data = [...classes];
      });
    } catch (error) {
      // TODO: implement error handling on UI
      console.error('Error handling not implemented');
      throw error;
    }
  }

  deleteClass(_id: string, name: string, level: string, numberOfStudents: number) {
    if (numberOfStudents > 0) {
      this.mswSnackbar.displayTimedMessage('לא ניתן למחוק את הכיתה כיוון שיש תלמידים המשוייכים אליה');
    } else {
      const dialogRef = this.dialog.open(DeleteClassDialogComponent, {
        data: { _id, name, level },
      });
      dialogRef.afterClosed().subscribe(async (result) => {
        if (result === true) {
          await this.classService.delete(_id);
        }
      });
    }
  }

  navigate(row_id: string) {
    this.router.navigate(['/class', row_id]);
  }
}
