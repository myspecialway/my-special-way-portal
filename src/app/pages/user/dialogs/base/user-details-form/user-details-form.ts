import { Component, OnInit, EventEmitter, Input, Output, OnDestroy } from '@angular/core';
import { FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
import { User, UserType } from '../../../../../models/user.model';
import { UserService } from '../../../services/user.service';
import { ClassService } from '../../../../class/services/class.graphql.service';
import { Class } from '../../../../../models/class.model';

export type UserTypeKey = keyof typeof UserType;

@Component({
  selector: 'app-user-details-form',
  templateUrl: './user-details-form.html',
  styleUrls: ['./user-details-form.scss'],
})
export class UserDetailsFormComponent implements OnInit, OnDestroy {
  form: FormGroup;
  roles: string[];
  userRoleEnum = UserType;
  classes: Class[];
  currentRole: UserTypeKey;
  formControl: FormControl;
  EmailFormControl: FormControl;
  selectUserType: FormControl;
  selectGrade: FormControl;
  userNameFormControl: FormControl;
  data: User;

  @Input()
  options = { submitButtonLabel: 'USER-DETAILS.ADD' };

  @Input('data')
  set _data(value: User) {
    this.data = value;
    this.currentRole = (value.role as any) as UserTypeKey;
  }

  @Output()
  formSubmit = new EventEmitter<User>();
  @Output()
  cancel = new EventEmitter<void>();

  constructor(private formBuilder: FormBuilder, private classService: ClassService, public userService: UserService) {
    this.roles = Object.keys(this.userRoleEnum);
  }

  ngOnInit(): void {
    this.attachClasses();
    this.createFormControls();
    this.createForm();
  }

  onUserTypeChange(userType: UserTypeKey): void {
    this.currentRole = userType;
    if (this.userRoleEnum[userType] === this.userRoleEnum.PRINCIPLE && this.data.class) {
      this.data.class = undefined;
    }
  }
  getClassDisabled() {
    return this.userRoleEnum[this.currentRole] !== this.userRoleEnum.TEACHER;
  }

  attachClasses() {
    this.classService.getAllClasses().subscribe((classes) => {
      this.classes = [...classes];
    });
  }

  close() {
    this.cancel.emit();
  }

  submitForm() {
    this.formSubmit.emit(this.form.value);
  }

  compareClass(c1: Class, c2: Class): boolean {
    return c1 && c2 ? c1._id === c2._id : c1 === c2;
  }

  private createFormControls() {
    this.formControl = new FormControl('', [Validators.required]);
    this.EmailFormControl = new FormControl('', [Validators.required, Validators.email]);
    this.selectUserType = new FormControl(null, Validators.required);
    this.selectGrade = new FormControl({ disabled: this.getClassDisabled() }, Validators.required);
    this.userNameFormControl = new FormControl('', [
      Validators.required,
      Validators.minLength(5),
      Validators.pattern('^[A-Za-z]+$'),
    ]);
  }

  private createForm() {
    this.form = this.formBuilder.group({
      firstname: '',
      lastname: '',
      username: this.userNameFormControl,
      email: this.EmailFormControl,
      role: this.selectUserType,
      class: this.selectGrade,
    });
  }

  ngOnDestroy() {}
}
