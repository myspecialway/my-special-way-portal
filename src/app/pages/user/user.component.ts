import { Component, ElementRef, OnInit, ViewChild, AfterViewInit } from '@angular/core';
import { MatDialog, MatSort, MatTableDataSource } from '@angular/material';
import { UserService } from './services/user.service';
import { User, UserType } from '../../models/user.model';
import { AddUserDialogComponent } from './dialogs/add/add-user.dialog';
import { DeleteUserDialogComponent } from './dialogs/delete/delete-user.dialog';
import * as _ from 'lodash';
import { UpdateUserDialogComponent } from './dialogs/update/update-user.dialog';

@Component({
  selector: 'app-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss'],
})
export class UserComponent implements OnInit, AfterViewInit {
  displayedColumns = ['name', 'username', 'class', 'type', 'enterPersonalArea', 'editDetails', 'deleteUser'];
  dataSource = new MatTableDataSource<User>();
  resultsLength = 0;

  @ViewChild(MatSort)
  sort: MatSort;
  @ViewChild('table')
  table: ElementRef;

  constructor(private userService: UserService, public dialog: MatDialog) {}

  ngOnInit(): void {
    this.userService.getAllUsers().subscribe((data) => {
      this.dataSource.data = data;
    });
  }

  ngAfterViewInit() {
    this.dataSource.sort = this.sort;
  }
  applyFilter(filterValue: string) {
    filterValue = filterValue.trim(); // Remove whitespace
    filterValue = filterValue.toLowerCase(); // Datasource defaults to lowercase matches
    this.dataSource.filter = filterValue;
  }

  toHebrew(type: UserType) {
    return UserType[type];
  }

  addNewUser() {
    const dialogRef = this.dialog.open(AddUserDialogComponent, {
      data: { user: User },
      height: '368px',
      width: '630px',
    });
    dialogRef.afterClosed().subscribe((data) => {
      if (data) {
        const newUser: User = this._createNewUser(data);
        this.userService.create(newUser);
      }
    });
  }
  deleteUser(_id: number, firstName: string, lastName: string, userType: UserType) {
    const dialogRef = this.dialog.open(DeleteUserDialogComponent, {
      data: { _id, firstName, lastName, userType },
      height: '275px',
      width: '360px',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result === true) {
        this.userService.delete(_id);
      }
    });
  }
  updateUser(_id: number, firstname: string, lastname: string, email: string, username: string, role: any, Class: any) {
    const dialogRef = this.dialog.open(UpdateUserDialogComponent, {
      data: { _id, firstname, lastname, email, username, role, Class },
      height: '376px',
      width: '631px',
    });

    dialogRef.afterClosed().subscribe((result) => {
      if (result) {
        const relevantUser = _.find(this.dataSource.data, { _id });
        const tempUser = _.assign({}, relevantUser, result);

        this.userService.update(tempUser);
      }
    });
  }

  _createNewUser(userData: any): User {
    const user: User = new User();
    user.firstname = userData.firstName;
    user.lastname = userData.lastName;
    user.username = userData.userName;
    user.email = userData.email;
    user.role = userData.userType;
    user.Class = userData.Class;
    return user;
  }
}
