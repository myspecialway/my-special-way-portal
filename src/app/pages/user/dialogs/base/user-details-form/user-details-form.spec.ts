import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/of';
import { classTestData } from './../../../../../../mocks/assets/classes.mock';
import { UserService } from './../../../services/user.service';
import { Apollo } from 'apollo-angular';
import { ClassService } from './../../../../class/services/class.graphql.service';
import { FormsModule, ReactiveFormsModule, FormBuilder } from '@angular/forms';
import { NO_ERRORS_SCHEMA } from '@angular/compiler/src/core';
import { UserDetailsFormComponent } from './user-details-form';
import { TestBed, ComponentFixture } from '@angular/core/testing';
import { User, UserType } from './../../../../../models/user.model';
import { Class } from '../../../../../models/class.model';
import { TranslateService, TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { TranslateCustomLoader } from '../../../../../../mocks/translate.stub';

describe('student component', () => {
  let dataMock: User;
  let fixture: ComponentFixture<UserDetailsFormComponent>;
  let component: UserDetailsFormComponent;
  let classServiceMock: Partial<ClassService>;

  beforeEach(async () => {
    classServiceMock = {
      getAllClasses: jest.fn().mockReturnValue(Observable.of(classTestData.classes)),
      delete: jest.fn(),
      create: jest.fn(),
      update: jest.fn(),
    };

    dataMock = {
      email: 'a',
      firstname: 'a',
      lastname: 'a',
      password: 'a',
      role: UserType.PRINCIPLE,
      username: 'a',
      _id: 0,
      class: new Class(),
    };

    TestBed.configureTestingModule({
      imports: [
        ReactiveFormsModule,
        FormsModule,
        TranslateModule.forRoot({
          loader: {
            provide: TranslateLoader,
            useClass: TranslateCustomLoader,
          },
        }),
      ],
      declarations: [UserDetailsFormComponent],
      providers: [FormBuilder, Apollo, UserService, { provide: ClassService, useValue: classServiceMock }],
      schemas: [NO_ERRORS_SCHEMA],
    });

    fixture = TestBed.createComponent(UserDetailsFormComponent);
    component = fixture.componentInstance;
    component.ngOnInit();
    // fixture.detectChanges();
  });

  it('should update current role and data on data @Input', () => {
    expect(component.currentRole).toBeUndefined();
    component._data = dataMock;
    // fixture.detectChanges();

    expect(component.data).toBe(dataMock);
    expect(component.currentRole).toBe(UserType.PRINCIPLE);
  });

  it('should update current role on onUserTypeChange', () => {
    component.onUserTypeChange('TEACHER');
    // fixture.detectChanges();

    expect(component.currentRole).toBe('TEACHER');
  });

  it('should reset Principle class on onUserTypeChange', () => {
    component._data = dataMock;
    expect(component.data.class).toBeInstanceOf(Class);

    component.onUserTypeChange('PRINCIPLE');
    // fixture.detectChanges();

    expect(component.data.class).toBeUndefined();
  });

  it('should emit cancel on close', () => {
    let called = false;
    component.cancel.subscribe(() => {
      called = true;
    });
    expect(called).toBeFalsy();

    component.close();

    // fixture.detectChanges();

    expect(called).toBeTruthy();
  });

  it('should trigger formSubmit form.value emission on submitForm', () => {
    let emittedValue = null;
    component.formSubmit.subscribe(() => {
      emittedValue = component.form.value;
    });
    expect(emittedValue).toBeNull();

    component.submitForm();

    expect(emittedValue).toEqual(component.form.value);
  });
});
