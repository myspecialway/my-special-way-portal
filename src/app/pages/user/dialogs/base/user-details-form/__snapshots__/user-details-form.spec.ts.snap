// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User component should render the component as described in snapshot 1`] = `
<app-user-details-form
  EmailFormControl={[Function FormControl]}
  authenticationService={[Function AuthenticationService]}
  cancel={[Function EventEmitter]}
  classService={[Function Object]}
  classes={[Function Array]}
  form={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  formControl={[Function FormControl]}
  formSubmit={[Function EventEmitter]}
  options={[Function Object]}
  roles={[Function Array]}
  selectClass={[Function FormControl]}
  selectUserType={[Function FormControl]}
  userNameFormControl={[Function FormControl]}
  userRoleEnum={[Function Object]}
  userService={[Function UserService]}
>
  <form
    novalidate=""
  >
    <mat-dialog-content>
      <div
        class="msw-div-column-right"
      >
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <input
              autocomplete="given-name"
              formcontrolname="firstname"
              matinput=""
              name="firstname"
              placeholder="שם פרטי"
              required=""
              tabindex="1"
            />
            
          </mat-form-field>
        </div>
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <input
              autocomplete="username"
              formcontrolname="username"
              matinput=""
              name="username"
              placeholder="שם משתמש"
              required=""
              tabindex="3"
            />
            
            
            
          </mat-form-field>
        </div>
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <mat-select
              autocomplete="role"
              formcontrolname="role"
              name="role"
              placeholder="סוג משתמש"
              required=""
              tabindex="5"
            >
              <mat-option
                disabled=""
              >
                בחר
              </mat-option>
              
            </mat-select>
            
          </mat-form-field>
        </div>
      </div>
      <div
        class="msw-div-column-left"
      >
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <input
              autocomplete="family-name"
              formcontrolname="lastname"
              matinput=""
              name="lastname"
              placeholder="שם משפחה"
              required=""
              tabindex="2"
            />
            
          </mat-form-field>
        </div>
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <input
              autocomplete="email"
              email=""
              formcontrolname="email"
              matinput=""
              name="email"
              placeholder="אימייל"
              required=""
              tabindex="4"
            />
            
            
          </mat-form-field>
        </div>
        <div
          class="msw-form-row"
        >
          <mat-form-field
            color="primary"
            hiderequiredmarker=""
          >
            <mat-select
              formcontrolname="class"
              name="class"
              placeholder="כיתה"
              tabindex="6"
            >
              <mat-option
                disabled=""
              >
                בחר
              </mat-option>
              
            </mat-select>
            
          </mat-form-field>
        </div>
      </div>
    </mat-dialog-content>
    <div
      mat-dialog-actions=""
    >
      <button
        class="msw-button-cancel"
        data-test-id="cancel-button"
        mat-button=""
        tabindex="-1"
        type="button"
      >
        ביטול
      </button>
      <button
        class="msw-button-submit"
        data-test-id="submit-button"
        mat-button=""
        type="submit"
      >
        
      </button>
    </div>
  </form>
</app-user-details-form>
`;
